libsnark/zk_proof_systems/plonk/plonk.hpp

```diff
@@ -0,0 +1,242 @@
+/** @file
+ *****************************************************************************
+ * @author     This file is part of libff, developed by Clearmatics Ltd
+ *             (originally developed by SCIPR Lab) and contributors
+ *             (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef LIBSNARK_ZK_PROOF_SYSTEMS_PLONK_PLONK_HPP_
+#define LIBSNARK_ZK_PROOF_SYSTEMS_PLONK_PLONK_HPP_
+
+#include <libff/algebra/curves/public_params.hpp>
+#include <libsnark/zk_proof_systems/plonk/tests/example.hpp>
+#include <memory>
+
+// enable debug checks. in particular enable comparisons to test
+// vector values.
+#define DEBUG
+//#undef DEBUG
+
+/// Declaration of common untility functions for ppzkSNARK proof
+/// system Plonk. The implementation instantiates the protocol of
+/// PlonK \[GWC19],
+///
+/// References:
+/// - \[GWC19]:
+///   Title: "Plonk: Permutations over lagrange-bases for oecumenical
+///   noninteractive arguments of knowledge", Ariel Gabizon, Zachary
+///   J. Williamson, and Oana Ciobotaru, Cryptology ePrint Archive,
+///   Report 2019/953, 2019, <https://eprint.iacr.org/2019/953>
+
+/// maximum degree of the encoded monomials in the usrs
+const size_t MAX_DEGREE = 254;
+
+/// number of generators for H, Hk1, Hk2
+const size_t NUM_HGEN = 3;
+
+namespace libsnark
+{
+
+enum W_polys_id { a, b, c };
+enum Q_polys_id { L, R, M, O, C };
+enum t_polys_id { lo, mid, hi };
+enum omega_id { base, base_k1, base_k2 };
+
+template<typename FieldT> using polynomial = std::vector<FieldT>;
```

Comment:
Hmm, I included the header but I get a compilation error that I am currently investigating:

`polynomial.hpp:17:51: error: ‘vector’ in namespace ‘std’ does not name a template type`

Reply:
OK.  So it sounds like `polynomial.hpp` depends on `#include <vector>`, but doesn't actually `#include` it.  That's my fault.  The fix is to include `<vector>` in `polynomial.hpp`.

Reply:
Fixed. Thanks!

Reply:
In general, it would be great if all include files bring in their dependencies.  It's actually a bit difficult to check this for every include, so mistakes can slip through.  But if we keep this approach in mind when we hit include issues then hopefully we can reduce the chances of this kind of problem in the future.

Reply:
Yes, I agree this is a very good rule. However I often have trouble enforcing it in a methodical way in practise. Let's maybe discuss this briefly in our next call.